plugins/cairo-lang-macro/src/lib.rs

```diff
@@ -143,6 +213,91 @@ impl AuxData {
     }
 }
 
+impl Diagnostic {
+    /// Convert to FFI-safe representation.
+    ///
+    /// # Safety
+    #[doc(hidden)]
+    pub fn into_stable(self) -> StableDiagnostic {
+        let cstr = CString::new(self.message).unwrap();
+        StableDiagnostic {
+            message: cstr.into_raw(),
+            severity: self.severity.into_stable(),
+        }
+    }
+
+    /// Convert to native Rust representation.
+    ///
+    /// # Safety
+    #[doc(hidden)]
+    pub unsafe fn from_stable(diagnostic: StableDiagnostic) -> Self {
+        Self {
+            message: raw_to_string(diagnostic.message),
+            severity: Severity::from_stable(diagnostic.severity),
+        }
+    }
+
+    /// Allocate dynamic array with FFI-safe diagnostics.
+    ///
+    /// # Safety
+    #[doc(hidden)]
+    pub unsafe fn allocate(diagnostics: Vec<Self>) -> (*mut StableDiagnostic, usize) {
+        let stable_diagnostics = diagnostics
+            .into_iter()
+            .map(|diagnostic| diagnostic.into_stable())
+            .collect::<Vec<_>>();
+        let n = stable_diagnostics.len();
+        let ptr = malloc(std::mem::size_of::<StableDiagnostic>() * n) as *mut StableDiagnostic;
```

Comment:
https://doc.rust-lang.org/stable/std/alloc/struct.System.html

> However, it is not valid to mix use of the backing system allocator with System, as this implementation may include extra work, such as to serve alignment requests greater than the alignment provided directly by the backing system allocator.

So I don't think it's valid to use `malloc/free` for the solution. Nor it is elegant IMHO.

Also, as far as I understand, you should be safe to use `Box<T, std::alloc::System>` in FFI boundary because [`Box` has a well-defined memory layout](https://doc.rust-lang.org/beta/std/boxed/index.html#memory-layout), and `std::alloc::System` should be always the same entity in our setup on all platforms.

And, because memory layout of `str` is the same as the layout of `[u8]`, I think you should be safe to use the following types, instead of pointer shenanigans that happen here:
* `Box<str, System>`
* `Box<[(Severity, Box<str, System>)], System>` (tuple might be replaced by `repr(C)` struct, I believe)

Because these types live within `System` allocator, their `Drop` implementations should deal with memory ownership for you, though you might need to use `std::mem::forget` and `Box::from_raw` or smth.

Reference: https://doc.rust-lang.org/reference/type-layout.html

wdyt?

Reply:
Thanks for linked materials! Seems like `Box<>` is the way to go indeed.